File: chrome/browser/webid/android/java/src/org/chromium/chrome/browser/webid/VerifiedOriginResolver.java
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

package org.chromium.chrome.browser.webid;

import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;

import androidx.browser.customtabs.CustomTabsService;

import org.jni_zero.CalledByNative;
import org.jni_zero.CalledByNativeForTesting;
import org.jni_zero.JNINamespace;
import org.jni_zero.JniType;
import org.jni_zero.NativeMethods;

import org.chromium.base.ContextUtils;
import org.chromium.base.Log;
import org.chromium.build.annotations.NullMarked;
import org.chromium.chrome.browser.browserservices.verification.ChromeOriginVerifier;
import org.chromium.url.Origin;

import java.util.List;

/** Resolves a FedCM Origin to a verified Android app package name and bound service name. */
@JNINamespace("content::webid")
@NullMarked
public class VerifiedOriginResolver {
    private static final String TAG = "OriginResolver";
    private static final String FEDCM_BOUND_SERVICE_INTENT_ACTION = "org.w3.FedCM";

    private long mNativeVerifiedOriginResolver;

    private VerifiedOriginResolver(long nativeVerifiedOriginResolver) {
        mNativeVerifiedOriginResolver = nativeVerifiedOriginResolver;
    }

    @CalledByNative
    private static VerifiedOriginResolver create(long nativeVerifiedOriginResolver) {
        return new VerifiedOriginResolver(nativeVerifiedOriginResolver);
    }

    @CalledByNative
    private void destroy() {
        mNativeVerifiedOriginResolver = 0;
    }

    @CalledByNative
    private void resolve(@JniType("url::Origin") Origin origin) {
        org.chromium.components.embedder_support.util.Origin dalOrigin =
                org.chromium.components.embedder_support.util.Origin.create(origin.toString());
        assert dalOrigin != null;

        Intent intent = new Intent(FEDCM_BOUND_SERVICE_INTENT_ACTION);
        Context context = ContextUtils.getApplicationContext();
        PackageManager packageManager = context.getPackageManager();
        List<ResolveInfo> services = packageManager.queryIntentServices(intent, 0);

        Log.d(TAG, "Discovered FedCM services: " + services);

        if (services.isEmpty()) {
            notifyResolved("", "");
            return;
        }

        List<String> packages = new java.util.ArrayList<>();
        for (ResolveInfo info : services) {
            packages.add(info.serviceInfo.packageName);
        }

        DigitalAssetLinksVerifier.checkPackages(
                packages,
                dalOrigin,
                index -> {
                    if (index != -1) {
                        ResolveInfo info = services.get(index);
                        notifyResolved(info.serviceInfo.packageName, info.serviceInfo.name);
                    } else {
                        Log.d(TAG, "No verified service found for origin");
                        notifyResolved("", "");
                    }
                });
    }

    private void notifyResolved(String packageName, String serviceName) {
        if (mNativeVerifiedOriginResolver != 0) {
            VerifiedOriginResolverJni.get()
                    .onOriginResolved(mNativeVerifiedOriginResolver, packageName, serviceName);
        }
    }

    @CalledByNativeForTesting
    private static void addVerificationOverrideForTesting(
            @JniType("std::string") String packageName, @JniType("url::Origin") Origin origin) {
        org.chromium.components.embedder_support.util.Origin dalOrigin =
                org.chromium.components.embedder_support.util.Origin.create(origin.toString());
        assert dalOrigin != null;
        ChromeOriginVerifier.addVerificationOverride(
                packageName, dalOrigin, CustomTabsService.RELATION_USE_AS_ORIGIN);
    }

    @NativeMethods
    interface Natives {
        void onOriginResolved(
                long nativeVerifiedOriginResolver,
                @JniType("std::string") String packageName,
                @JniType("std::string") String serviceName);
    }
}

File: chrome/browser/webid/verified_origin_resolver.cc
// Copyright 2026 The Chromium Authors
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include "chrome/browser/webid/verified_origin_resolver.h"

#include "base/android/jni_android.h"
#include "base/android/jni_string.h"
#include "chrome/browser/webid/jni_headers/VerifiedOriginResolver_jni.h"
#include "url/origin.h"

namespace content::webid {
VerifiedOriginResolver::VerifiedOriginResolver() {
  JNIEnv* env = base::android::AttachCurrentThread();
  java_obj_.Reset(Java_VerifiedOriginResolver_create(
      env, reinterpret_cast<intptr_t>(this)));
}

VerifiedOriginResolver::~VerifiedOriginResolver() {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_VerifiedOriginResolver_destroy(env, java_obj_);
}

void VerifiedOriginResolver::Resolve(const url::Origin& origin,
                                     ResolveCallback callback) {
  JNIEnv* env = base::android::AttachCurrentThread();
  callback_ = std::move(callback);
  Java_VerifiedOriginResolver_resolve(env, java_obj_, origin);
}

// static
void VerifiedOriginResolver::AddVerificationOverrideForTesting(  // IN-TEST
    const std::string& package_name,
    const url::Origin& origin) {
  JNIEnv* env = base::android::AttachCurrentThread();
  Java_VerifiedOriginResolver_addVerificationOverrideForTesting(  // IN-TEST
      env, package_name, origin);
}

void VerifiedOriginResolver::OnOriginResolved(JNIEnv* env,
                                              const std::string& package_name,
                                              const std::string& service_name) {
  if (!callback_) {
    return;
  }
  if (package_name.empty() || service_name.empty()) {
    std::move(callback_).Run(base::unexpected(ResolveError::kNoServiceFound));
  } else {
    std::move(callback_).Run(std::make_pair(package_name, service_name));
  }
}

}  // namespace content::webid

DEFINE_JNI_FOR_VerifiedOriginResolver()  // NOLINT(readability/fn_size)
